src/main/java/de/tum/in/www1/artemis/service/FileService.java

```diff
@@ -101,6 +129,84 @@ public void resetOnPath(String path) {
         // Intentionally blank
     }
 
+    /**
+     * Helper method which handles the file creation for both normal file uploads and for markdown
+     *
+     * @param file         The file to be uploaded with a maximum file size set in resources/config/application.yml
+     * @param keepFileName specifies if original file name should be kept
+     * @param markdown     boolean which is set to true, when we are uploading a file within the markdown editor
+     * @return The path of the file
+     */
+    @NotNull
+    public String handleSaveFile(MultipartFile file, boolean keepFileName, boolean markdown) {
+        // check for file type
+        String filename = file.getOriginalFilename();
+        if (filename == null) {
+            throw new IllegalArgumentException("Filename cannot be null");
+        }
+        // sanitize the filename and replace all invalid characters with "_"
+        filename = filename.replaceAll("[^a-zA-Z\\d\\.\\-]", "_");
+        String fileExtension = FilenameUtils.getExtension(filename);
+        if (this.allowedFileExtensions.stream().noneMatch(fileExtension::equalsIgnoreCase)) {
+            throw new BadRequestException("Unsupported file type! Allowed file types: " + String.join(", ", this.allowedFileExtensions));
+        }
+
+        final String filePath = markdown ? FilePathService.getMarkdownFilePath() : FilePathService.getTempFilePath();
+        final String fileNameAddition = markdown ? "Markdown_" : "Temp_";
+        final StringBuilder responsePath = new StringBuilder(markdown ? MARKDOWN_FILE_SUBPATH : DEFAULT_FILE_SUBPATH);
+
+        try {
+            File newFile = createNewFile(filePath, filename, fileNameAddition, fileExtension, keepFileName);
+            responsePath.append(newFile.toPath().getFileName());
+
+            // copy contents of uploaded file into newly created file
+            Files.copy(file.getInputStream(), newFile.toPath(), REPLACE_EXISTING);
+
+            return responsePath.toString();
+        }
+        catch (IOException e) {
+            log.error("Could not save file {}", filename, e);
+            throw new InternalServerErrorException("Could not create file");
+        }
+    }
+
+    /**
+     * Creates a new file from given contents
+     * @param filePath the path to save the file to excluding the filename
+     * @param filename the filename of the file to save
+     * @param fileNameAddition the addition to the filename to make sure it is unique
+     * @param fileExtension the extension of the file to save
+     * @param keepFileName specifies if original file name should be kept
+     * @return the created file
+     */
+    private File createNewFile(String filePath, String filename, String fileNameAddition, String fileExtension, boolean keepFileName) throws IOException {
+        try {
+            Files.createDirectory(Paths.get(filePath));
```

Comment:
```suggestion
            Files.createDirectories(Paths.get(filePath));
```

There is an important difference between the two methods here:

> Creates a directory by creating all nonexistent parent directories first. Unlike the [createDirectory](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/nio/file/Files.html#createDirectory(java.nio.file.Path,java.nio.file.attribute.FileAttribute...)) method, an exception is not thrown if the directory could not be created because it already exists.

An existing directory should not be a problem here, I think?

Reply:
Good point